pkg/triggers/lease_backend.go

```diff
@@ -0,0 +1,44 @@
+package triggers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// LeaseBackend does a check and set operation on the Lease object in the defined data source
+//
+//go:generate mockgen -destination=./mock_lease_backend.go -package=triggers "github.com/kubeshop/testkube/pkg/triggers" LeaseBackend
+type LeaseBackend interface {
+	// CheckAndSet should do a check and set operation with upsert
+	CheckAndSet(ctx context.Context, id string) (*Lease, error)
+}
+
+type MongoLeaseBackend struct {
+	coll *mongo.Collection
+}
+
+func NewMongoLeaseBackend(db *mongo.Database) *MongoLeaseBackend {
+	return &MongoLeaseBackend{coll: db.Collection(mongoCollectionTriggersLease)}
+}
+
+func (b *MongoLeaseBackend) CheckAndSet(ctx context.Context, id string) (*Lease, error) {
+	after := options.After
+	upsert := true
+	opts := options.FindOneAndUpdateOptions{
+		ReturnDocument: &after,
+		Upsert:         &upsert,
+	}
+	res := b.coll.FindOneAndUpdate(ctx, bson.M{"_id": _id, "ds": ""}, bson.M{}, &opts)
```

Comment:
And for multiple api replicas, we will update it multiple times?

Reply:
Lease logic does the following:
* Get current lease
* if it doesn't exist, create it, if it fails because of duplicate key, somebody created it in the meantime and got the lease, the current replica will not get the lease
* if it exists, try to do a findAndUpdate which is an atomic operation in Mongo
* check the identifier in the returned document from findAndUpdate operation, if other replica did it in the meantime

Reply:
Yes, but it might fail, if a few replcas try to obtain the lease for all of them. From what I remember, it was suggested to use retry appoach, at least we do it for creating uniqiue numbers for executions

Reply:
I am not sure I fully understand the idea, can we discuss it internally?